tools/integration_tests/operations/list_dir_test.go

```diff
@@ -0,0 +1,143 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Provides integration tests for list directory.
+package operations_test
+
+import (
+	"io/fs"
+	"log"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/googlecloudplatform/gcsfuse/tools/integration_tests/util/setup"
+)
+
+func createDirectoryWithFile(dirPath string, filePath string, t *testing.T) {
+	err := os.Mkdir(dirPath, setup.FilePermission_0600)
+	if err != nil {
+		t.Errorf("Mkdir at %q: %v", dirPath, err)
+		return
+	}
+
+	_, err = os.Create(filePath)
+	if err != nil {
+		t.Errorf("Error in creating file %v:", err)
+	}
+}
+
+func checkIfListedCorrectDirectory(dirPath string, obj fs.DirEntry, t *testing.T) {
+	var dirName string
+
+	if strings.Contains(dirPath, SecondSubDirectoryForListTest) {
+		dirName = SecondSubDirectoryForListTest
+	} else if strings.Contains(dirPath, FirstSubDirectoryForListTest) {
+		dirName = FirstSubDirectoryForListTest
+	} else if strings.Contains(dirPath, DirectoryForListTest) {
+		dirName = DirectoryForListTest
+	} else if strings.Contains(dirPath, setup.MntDir()) {
+		dirName = setup.MntDir()
+	}
+
+	switch dirName {
+	case setup.MntDir():
+		{
+			// testBucket/directoryForListTest    -- Dir
+			if obj.Name() != DirectoryForListTest || obj.IsDir() != true {
+				t.Errorf("Listed incorrect object.")
+			}
+		}
+	case DirectoryForListTest:
+		{
+			// testBucket/directoryForListTest/fileInDirectoryForListTest     -- File
+			// testBucket/directoryForListTest/firstSubDirectoryForListTest   -- Dir
+			// testBucket/directoryForListTest/secondSubDirectoryForListTest  -- Dir
+			if (obj.Name() != FileInDirectoryForListTest && obj.IsDir() == true) && (obj.Name() != FirstSubDirectoryForListTest && obj.IsDir() != true) && (obj.Name() != SecondSubDirectoryForListTest && obj.IsDir() != true) {
+				t.Errorf("Listed incorrect object")
+			}
+		}
+	case FirstSubDirectoryForListTest:
+		{
+			// testBucket/directoryForListTest/firstSubDirectoryForListTest/fileInFirstSubDirectoryForListTest     -- File
+			if obj.Name() != FileInFirstSubDirectoryForListTest && obj.IsDir() == true {
+				t.Errorf("Listed incorrect object")
+			}
+		}
+	case SecondSubDirectoryForListTest:
+		{
+			// testBucket/directoryForListTest/secondSubDirectoryForListTest/fileInSecondSubDirectoryForListTest   -- File
+			if obj.Name() != FileInSecondSubDirectoryForListTest && obj.IsDir() == true {
+				t.Errorf("Listed incorrect object")
+			}
+		}
+	}
+}
+
+// List directory recursively
+func listDirectory(path string, t *testing.T) {
+	//Reading contents of the directory
+	objs, err := os.ReadDir(path)
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	for _, obj := range objs {
+		checkIfListedCorrectDirectory(path, obj, t)
```

Comment:
This checks if the listed object is under correct directory or not. However, let's say there is a bug in code and some object in a directory doesn't appear in "objs" itself then this test case will still pass. Can we cover that case as well ?